utils/service/txmgr/txmgr.go

```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
-		} else {
-			m.l.Debug("Overriding the fee cap to enforce a price bump", "previous", tx.GasFeeCap(), "suggested", gasFeeCap, "new", thresholdFeeCap)
-		}
-		gasFeeCap = thresholdFeeCap
-	}
-
-	if reusedTip && reusedFeeCap {
-		return tx, nil
-	}
-
-	rawTx := &types.DynamicFeeTx{
-		ChainID:    tx.ChainId(),
-		Nonce:      tx.Nonce(),
-		GasTipCap:  gasTipCap,
-		GasFeeCap:  gasFeeCap,
-		Gas:        tx.Gas(),
-		To:         tx.To(),
-		Value:      tx.Value(),
-		Data:       tx.Data(),
-		AccessList: tx.AccessList(),
-	}
-	return m.Signer(ctx, m.From, types.NewTx(rawTx))
+	metr    metrics.TxMetricer
 }
 
 // NewSimpleTxManager initializes a new SimpleTxManager with the passed Config.
-func NewSimpleTxManager(name string, l log.Logger, cfg Config, backend ETHBackend) *SimpleTxManager {
-	if cfg.NumConfirmations == 0 {
-		panic("txmgr: NumConfirmations cannot be zero")
+func NewSimpleTxManager(name string, l log.Logger, m metrics.TxMetricer, cfg CLIConfig) (*SimpleTxManager, error) {
+	conf, err := NewConfig(cfg, l)
+	if err != nil {
+		return nil, err
 	}
 
 	return &SimpleTxManager{
+		chainID: conf.ChainID,
 		name:    name,
-		Config:  cfg,
-		backend: backend,
+		cfg:     conf,
+		backend: conf.Backend,
 		l:       l.New("service", name),
-	}
+		metr:    m,
+	}, nil
+}
+
+func (m *SimpleTxManager) From() common.Address {
+	return m.cfg.From
+}
+
+// TxCandidate is a transaction candidate that can be submitted to ask the
+// [TxManager] to construct a transaction with gas price bounds.
+type TxCandidate struct {
+	// TxData is the transaction data to be used in the constructed tx.
+	TxData []byte
+	// To is the recipient of the constructed tx. Nil means contract creation.
+	To *common.Address
+	// GasLimit is the gas limit to be used in the constructed tx.
+	GasLimit uint64
 }
 
 // Send is used to publish a transaction with incrementally higher gas prices
 // until the transaction eventually confirms. This method blocks until an
 // invocation of sendTx returns (called with differing gas prices). The method
 // may be canceled using the passed context.
 //
-// The initially supplied transaction must be signed, have gas estimation done, and have a reasonable gas fee.
-// When the transaction is resubmitted the tx manager will re-sign the transaction at a different gas pricing
-// but retain the gas used, the nonce, and the data.
+// The transaction manager handles all signing. If and only if the gas limit is 0, the
+// transaction manager will do a gas estimation.
 //
 // NOTE: Send should be called by AT MOST one caller at a time.
-func (m *SimpleTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+func (m *SimpleTxManager) Send(ctx context.Context, candidate TxCandidate) (*types.Receipt, error) {
+	if m.cfg.TxSendTimeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, m.cfg.TxSendTimeout)
+		defer cancel()
```

Comment:
It looks `var cancle` should be moved up!

Reply:
I don't know what you mean. Where to move the line `var cancel context.CancelFunc`?

Reply:
Oh, i thought this `defer cancel()` is executed too early that it cancel first before executing any. But it is not! Just forget about my comment

```go
package main

import "fmt"

func main() {
	flag := true
	if flag {
		defer fmt.Println("hello inside the if")
	}

	fmt.Println("hello")
}
```

```shell
> go run .
hello
hello inside the if
```